### PR TITLE
Added Qwen2-MoE model, optimizing its performance on Gaudi

### PR DESCRIPTION
**Pull request by gyou2021**
… with kv_cache, reuse_cache, flash attention, internal bucket, etc.

# What does this PR do?
Added Qwen2-MoE model into optimum-habana, optimizing its performance on Gaudi with kv_cache, reuse_cache, flash attention, internal bucket, etc.

Test command:
python run_generation.py --model_name_or_path <model path> --use_hpu_graphs --use_kv_cache --reuse_cache --trim_logits --attn_softmax_bf16 --use_flash_attention --max_input_tokens 256 --max_new_tokens 256 --batch_size 2 --limit_hpu_graphs --bucket_size 128 --bucket_internal --n_iterations 2 --warmup 2 --bf16

### CODE REVIEW

File: optimum/habana/transformers/models/qwen2_moe/modeling_qwen2_moe.py [hunk 1]
```diff
@@ -0,0 +1,1144 @@
...
+                #     and num_virtual_tokens is not None
+                #     and num_virtual_tokens == past_key_value[0].shape[-2]
+                # ):
+                #     # prefix tunining case. attach past_key_value to generate first token.
+                #     key_states = torch.cat((past_key_value[0], key_states), -2)
+                #     value_states = torch.cat((past_key_value[1], value_states), -2)
+                #     past_key_value = (key_states, value_states)
+                # else:
+                #     key_states = self.k_cache.update(past_key_value[0], key_states, 2, token_idx, self.inp_seq_len)
+                #     value_states = self.v_cache.update(past_key_value[1], value_states, 2, token_idx, self.inp_seq_len)
```
Comment (ssarkar2) [inline optimum/habana/transformers/models/qwen2_moe/modeling_qwen2_moe.py:395]: clean up

File: optimum/habana/transformers/models/qwen2_moe/modeling_qwen2_moe.py [hunk 2]
```diff
@@ -0,0 +1,1144 @@
...
+        query_states, key_states = apply_customized_rope(query_states, key_states, cos, sin, position_ids)
+        #query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
+               
+        if use_cache:
+            # reuse k, v, self_attention
+            if reuse_cache:
+                # if past_key_value is not None and isinstance(past_key_value[0], torch.Tensor):
+                #     # prefix tuning case. attach past_key_value to generate first token.
+                #     key_states = torch.cat((past_key_value[0], key_states), -2)
+                #     value_states = torch.cat((past_key_value[1], value_states), -2)
```
Comment (ssarkar2) [inline optimum/habana/transformers/models/qwen2_moe/modeling_qwen2_moe.py:369]: clean up

**Review by ssarkar2 — COMMENTED** [review]
please add a short test here:
https://github.com/huggingface/optimum-habana/blob/main/tests/test_text_generation_example.py

**Comment by emascarenhas** [comment]
Please sync your PR with main/upstream and fix any merge conflicts. Thank you.

**Comment by gyou2021** [comment]
> please add a short test here: https://github.com/huggingface/optimum-habana/blob/main/tests/test_text_generation_example.py

Added a Qwen2-MoE short test in https://github.com/gyou2021/optimum-habana/blob/main/tests/test_text_generation_example.py

**Comment by gyou2021** [comment]
> Please sync your PR with main/upstream and fix any merge conflicts. Thank you.

Synced with main/upstream and fixed merge conflicts.

**Comment by gyou2021** [comment]
I created a new branch (https://github.com/huggingface/optimum-habana/pull/1316 ) following the contributor guidance. The PR 1162 was changed to the PR 1316.
